### PR TITLE
Add profiling to dataloader

### PR DESCRIPTION
**Pull request by akashkw**
## What does this PR do?

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 1]
```diff
@@ -107,7 +107,8 @@ def advance(  # type: ignore[override]
         void(dl_max_batches)
 
         assert self._dataloader_iter is not None
-        with self.trainer.profiler.profile("{}_dataloader_next"):
+        action_name = "test_dataloader_next" if self.trainer.testing else "val_dataloader_next"
```
Comment (ananthsub) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:110]: let's follow the action name style here so it's uniform in the profiler later
https://github.com/PyTorchLightning/pytorch-lightning/blob/7e2f9fbad555242b0ceb2a24e5e4c004f0701bae/pytorch_lightning/trainer/trainer.py#L1604

        with self.profiler.profile(f"[LightningModule]{pl_module.__class__.__name__}.{hook_name}"):

```suggestion
        suffix = "test_dataloader_next" if self.trainer.testing else "val_dataloader_next"
        action_name = f"[{self.__class__.__name__}].{suffix}"
```

Comment (akashkw) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py]: There are a couple spots that don't follow this convention. Should these be updated too?

- https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loops/fit_loop.py#L273
- https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loops/epoch/training_epoch_loop.py#L198

**Review by carmocca — CHANGES_REQUESTED** [review]
Hi @akashkw!

We cannot profile `next()` like this as this would not cover prefetched batches or inter-batch paralelism.

I recently had a call with @daniellepintz where I went over possible solutions. Here's some code pointers:

The concrete lines to profile:
https://github.com/PyTorchLightning/pytorch-lightning/blob/a0655611de460f1659b46150cda256d2e3fa974e/pytorch_lightning/utilities/fetching.py#L267


https://github.com/PyTorchLightning/pytorch-lightning/blob/a0655611de460f1659b46150cda256d2e3fa974e/pytorch_lightning/utilities/fetching.py#L336

Option 1:
Pass a profiler reference to the data fetchers, maybe also the stage and dataloader idx for more fine-grained profiling, then use these directly inside the fetching functions
Option 2:
Inject the profiler usage through callables that could be part of these hooks:
https://github.com/PyTorchLightning/pytorch-lightning/blob/a0655611de460f1659b46150cda256d2e3fa974e/pytorch_lightning/utilities/fetching.py#L61-L65
with a pattern similar to the one used in this PR to inject logic around the optimizer.step()
https://github.com/PyTorchLightning/pytorch-lightning/commit/4cc05b2cf98e49168a5f5dc265647d75d1d3aae9

**Comment by akashkw** [comment]
> Option 1:
> Pass a profiler reference to the data fetchers, maybe also the stage and dataloader idx for more fine-grained profiling, then  use these directly inside the fetching functions
> Option 2:
> Inject the profiler usage through callables that could be part of these hooks:

@carmocca do you have any preference between these two approaches?

**Comment by carmocca** [comment]
I prefer option (2) so the fetching does not need to keep a reference to the trainer. (the profiler is owned by the trainer)

File: pytorch_lightning/loops/epoch/prediction_epoch_loop.py [hunk 2]
```diff
@@ -89,7 +89,8 @@ def advance(  # type: ignore[override]
             num_dataloaders: the total number of dataloaders
             return_predictions: whether to return the obtained predictions
         """
-        batch_idx, batch = next(dataloader_iter)
+        with self.trainer.profiler.profile(f"[{self.__class__.__name__}].pred_dataloader_next"):
+            batch_idx, batch = next(dataloader_iter)
```
Comment (akashkw) [inline pytorch_lightning/loops/epoch/prediction_epoch_loop.py:93]: `prediction_epoch_loop` doesn't seem to use data fetchers, should I just leave this as is?

Comment (carmocca) [reply pytorch_lightning/loops/epoch/prediction_epoch_loop.py:93]: Yes

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 3]
```diff
@@ -88,6 +88,11 @@ def on_run_start(  # type: ignore[override]
         # add the previous `fetched` value to properly track `is_last_batch` with no prefetching
         data_fetcher.fetched += self.batch_progress.current.ready
 
+        data_fetcher._start_profiler = self.trainer.profiler.start
+        data_fetcher._stop_profiler = self.trainer.profiler.stop
+        suffix = "test_dataloader_next" if self.trainer.testing else "val_dataloader_next"
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:93]: I think we should differentiate per dataloader idx too

File: pytorch_lightning/loops/epoch/prediction_epoch_loop.py [hunk 4]
```diff
@@ -89,7 +89,8 @@ def advance(  # type: ignore[override]
             num_dataloaders: the total number of dataloaders
             return_predictions: whether to return the obtained predictions
         """
-        batch_idx, batch = next(dataloader_iter)
+        with self.trainer.profiler.profile(f"[{self.__class__.__name__}].pred_dataloader_next"):
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/prediction_epoch_loop.py:92]:
```suggestion
        with self.trainer.profiler.profile(f"[{self.__class__.__name__}].predict_dataloader_next"):
```

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 5]
```diff
@@ -88,6 +88,11 @@ def on_run_start(  # type: ignore[override]
...
+        data_fetcher._profiler_action_name = f"[{self.__class__.__name__}].{suffix}"
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:94]: Let's avoid the `_profiler_action_name` attribute in the fetcher

```python
data_fetcher._start_profiler = self._on_before_fetch

def on_before_fetch(self):
    suffix = "test_dataloader_next" if self.trainer.testing else "val_dataloader_next"
    self.trainer.profiler.start(f"[{self.__class__.__name__}].{suffix}")
```


You could also use `functools.partial` but I would say this is nicer

File: pytorch_lightning/utilities/fetching.py [hunk 6]
```diff
@@ -385,3 +398,9 @@ def fetching_function(self) -> Tuple[int, Iterator]:
         if not self.done:
             return self.fetched, self.iterator
         raise StopIteration
+
+    def on_fetch_start(self) -> None:
```
Comment (carmocca) [inline pytorch_lightning/utilities/fetching.py:402]: These are only called in `DataFetcher._fetch_next_batch` but that method is not used by this class (this class subclasses `AbstractDataFetcher`)

Particularly, this start and stop needs to wrap the `next` in `StepFuncDataLoaderIter`

**Review by carmocca — COMMENTED** [review]
Good progress!

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 5 above)
Comment (akashkw) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:94]: Didn't know about `functools.partial`, thanks for sharing!
I agree, this is approach is cleaner.

File: pytorch_lightning/utilities/fetching.py [hunk 7]
```diff
@@ -320,9 +330,11 @@ def move_to_device(self, batch: Any) -> Any:
 
     def on_fetch_start(self) -> "torch.cuda.Event":
         # create a cuda event used to record the async stream of data to device.
+        self._start_profiler()
         return torch.cuda.Event()
```
Comment (carmocca) [inline pytorch_lightning/utilities/fetching.py:334]: Probably overkill but if creating the `cuda.Event()` was a very expensive call, then we would want to do

```suggestion
        event = torch.cuda.Event()
        self._start_profiler()
        return event
```

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 8]
```diff
@@ -146,6 +146,15 @@ def on_run_start(self, data_fetcher: AbstractDataFetcher) -> None:  # type: igno
         # add the previous `fetched` value to properly track `is_last_batch` with no prefetching
         data_fetcher.fetched += self.batch_progress.current.ready
 
+        data_fetcher._start_profiler = self._on_before_fetch
+        data_fetcher._stop_profiler = self._on_after_fetch
+
+    def _on_before_fetch(self):
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:152]: Remember typing (see mypy failures in CI)

**Review by carmocca — COMMENTED** [review]
You should check for these new entries in the profiler tests

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 9]
```diff
@@ -88,6 +88,21 @@ def on_run_start(  # type: ignore[override]
         # add the previous `fetched` value to properly track `is_last_batch` with no prefetching
         data_fetcher.fetched += self.batch_progress.current.ready
 
+        dataloader_idx = f"_{kwargs['dataloader_idx']}" if "dataloader_idx" in kwargs else ""
+        if self.trainer.testing:
+            suffix = f"test_dataloader{dataloader_idx}_next"
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:93]: In result logging, we use the convention `dataloader_idx_{}`. We should do the same here too

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 10]
```diff
@@ -88,6 +88,21 @@ def on_run_start(  # type: ignore[override]
...
+        else:
+            suffix = f"val_dataloader{dataloader_idx}_next"
+
+        def _on_before_fetch() -> None:
+            self.trainer.profiler.start(f"[{self.__class__.__name__}].{suffix}")
+
+        def _on_after_fetch() -> None:
+            self.trainer.profiler.stop(f"[{self.__class__.__name__}].{suffix}")
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:101]: I'm uncomfortable using local functions as they will not work if this class gets pickled.

**Review by carmocca — COMMENTED** [review]
CHANGELOG and tests!

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:101]: Same here. Let's avoid them.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 11]
```diff
@@ -146,6 +146,15 @@ def on_run_start(self, data_fetcher: AbstractDataFetcher) -> None:  # type: igno
         # add the previous `fetched` value to properly track `is_last_batch` with no prefetching
         data_fetcher.fetched += self.batch_progress.current.ready
 
+        def _on_before_fetch() -> None:
+            self.trainer.profiler.start(f"[{self.__class__.__name__}].train_dataloader_next")
```
Comment (tchaton) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:150]: Same here.

**Review by carmocca — APPROVED** [review]

File: tests/utilities/test_fetching.py [hunk 12]
```diff
@@ -485,3 +486,88 @@ def validation_step(self, batch, batch_idx):
     assert dm.count_called_on_before_batch_transfer == 4
     assert dm.count_called_transfer_batch_to_device == 4
     assert dm.count_called_on_after_batch_transfer == 4
+
+
+@RunIf(skip_windows=True)  # TODO: all durations are 0 on Windows
```
Comment (carmocca) [inline tests/utilities/test_fetching.py:491]: I'm unsure if this is expected or a bug...

**Review by rohitgr7 — APPROVED** [review]

**Review by otaj — APPROVED** [review]
